### PR TITLE
Subqueries in JOIN

### PR DESCRIPTION
**Pull request by btry**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none

This PR is a proposal to allows to create subqueries in JOIN. Example in the unit tests.

Example of the need in Flyve MDM: https://github.com/flyve-mdm/glpi-plugin/pull/785

```php
      $result = $DB->request("SELECT `l1`.*
         FROM `$table`  AS `l1` INNER JOIN
         (SELECT  MAX(`date`) AS `maxdate`, `topic`
            FROM `$table`
            WHERE `itemtype` = '$itemtype' AND `items_id` = '$itemId'
            GROUP BY `topic`
         ) AS `l2`
         ON `l1`.`topic` = `l2`.`topic` AND `l1`.`date` = `l2`.`maxdate`
         WHERE `itemtype`='$itemtype' AND `items_id` = '$itemId'
         GROUP BY `topic`");
```

In this example I need to get a single row for the column topic, with the latest associated value. Without a subquery, mysql messes the rows and may give a value of the column message for an date which is not the highest. Using a subquery like this one is requires to have accurate results.

### CODE REVIEW

File: inc/dbmysqliterator.class.php [hunk 1]
```diff
@@ -194,12 +194,10 @@ function buildQuery ($table, $crit = "", $log = false) {
                      break;
 
                   case 'JOIN' : // deprecated
+                     Toolbox::deprecated('"JOIN" is deprecated, please use "LEFT JOIN" instead.');
```
Comment (trasher) [inline inc/dbmysqliterator.class.php]: I do not like this way; and as far as I remember this already cause difficult to understand issues.

Comment (trasher) [inline inc/dbmysqliterator.class.php]: I do not like this way; and as far as I remember this already cause difficult to understand issues.

Comment (trasher) [inline inc/dbmysqliterator.class.php]: I do not like this way; and as far as I remember this already cause difficult to understand issues.

**Comment by trasher** [comment]
Not able to add a comment directly in the code... For the JOIN depreacated, I do not like this way; and as far as I remember this already cause difficult to understand issues. Please leave as is in 9.3.

File: inc/dbmysqliterator.class.php (hunk 1 above)
Comment (trasher) [inline inc/dbmysqliterator.class.php]: I do not like this way; and as far as I remember this already cause difficult to understand issues. Please leave as is in 9.3.

**Comment by btry** [comment]
I cleaned up duplicates of your comment.

**Comment by btry** [comment]
I cleaned up duplicates of your comment; but there is something weird today.... the last one gone as well :/

File: inc/dbmysqliterator.class.php [hunk 2]
```diff
@@ -215,7 +215,16 @@ function buildQuery ($table, $crit = "", $log = false) {
                               break;
                         }
                         foreach ($val as $jointable => $joincrit) {
-                           $join .= " $jointype JOIN " .  DBmysql::quoteName($jointable) . " ON (" . $this->analyseCrit($joincrit) . ")";
+                           $joinExpressionKeys = ['FKEY' => null, 'ON' => null];
```
Comment (btry) [inline inc/dbmysqliterator.class.php]: Might be an opportuniy to have  this array in a private property and mutualize it with this liine 
https://github.com/glpi-project/glpi/blob/9.3/bugfixes/inc/dbmysqliterator.class.php#L490

**Comment by btry** [comment]
I updated the PR. Removed the refactor, and added an extra commit to not have the subquery and the FKEY in the same array. It looks smarter to me.

@trasher 

As dusussed, this is compatible with Postgres as it allows a derived table to be used for JOIN, at least since 7.2

File: inc/dbmysqliterator.class.php (hunk 2 above)
Comment (trasher) [reply inc/dbmysqliterator.class.php]: This should be factorized, yes.

I'm do not know yet if I prefer your prévious proposal or this one :D But I'm globally OK with the change, thank you.

Comment (btry) [reply inc/dbmysqliterator.class.php]: the new one avoids the ugly key "SUBQUERY". This is smoother to write, and seems to me more natural and easier to learn.

Comment (trasher) [reply inc/dbmysqliterator.class.php]: Building queries from an array is not natural :D

**Comment by trasher** [comment]
Could you please rebase your branch on master? Thank you :)

**Comment by btry** [comment]
@trasher : done

**Pull request merged**